File: api_bus/core/models/boleto.py
from django.db import models
from .usuario import Usuario
from .ruta import Ruta
from .autobus import Autobus
from .horario import Horario

class Boleto(models.Model):
    ESTADO_CHOICES = [
        ('reservado', 'Reservado'),
        ('cancelado', 'Cancelado'),
        ('completado', 'Completado'),
    ]
    
    usuario = models.ForeignKey(Usuario, on_delete=models.CASCADE, related_name='boletos')
    ruta = models.ForeignKey(Ruta, on_delete=models.CASCADE)
    autobus = models.ForeignKey(Autobus, on_delete=models.CASCADE)
    horario = models.ForeignKey(Horario, on_delete=models.CASCADE)
    fecha_reservacion = models.DateTimeField(auto_now_add=True)
    fecha_viaje = models.DateField()
    asiento_numero = models.PositiveIntegerField()
    estado = models.CharField(max_length=10, choices=ESTADO_CHOICES, default='reservado')
    precio = models.DecimalField(max_digits=10, decimal_places=2)
    
    class Meta:
        unique_together = ('autobus', 'fecha_viaje', 'asiento_numero', 'horario')
    
    def __str__(self):
        return f"Boleto #{self.id} - {self.usuario.persona.nombre}"

File: api-v2/models/boleto.js
// models/boleto.js (ESM)
import { DataTypes } from 'sequelize';
import sequelize from '../config/db.js';

const ESTADO = ['reservado', 'cancelado', 'completado'];

const Boleto = sequelize.define(
  'Boleto',
  {
    id: {
      type: DataTypes.BIGINT.UNSIGNED,
      autoIncrement: true,
      primaryKey: true,
    },
    usuarioId: {
      type: DataTypes.BIGINT.UNSIGNED,
      allowNull: true,
        onUpdate: 'CASCADE',
      onDelete: 'SET NULL',
      field: 'usuario_id',
    },
    // nuevo campo para almacenar nombre del usuario
    nombre_usuario: {
      type: DataTypes.STRING(255),
      allowNull: true,
    },
    rutaId: {
      type: DataTypes.BIGINT.UNSIGNED,
      allowNull: true,
      references: { model: 'Rutas', key: 'id' },
      onUpdate: 'CASCADE',
      onDelete: 'SET NULL',
      field: 'ruta_id',
    },
    autobusId: {
      type: DataTypes.BIGINT.UNSIGNED,
      allowNull: true,
      references: { model: 'Autobuses', key: 'id' },
      onUpdate: 'CASCADE',
      onDelete: 'SET NULL',
      field: 'autobus_id',
    },
    horarioId: {
      type: DataTypes.BIGINT.UNSIGNED,
      allowNull: true,
      references: { model: 'Horarios', key: 'id' },
      onUpdate: 'CASCADE',
      onDelete: 'SET NULL',
      field: 'horario_id',
    },
    fecha_reservacion: {
      type: DataTypes.DATE,
      allowNull: true,
      defaultValue: DataTypes.NOW,
    },
    fecha_viaje: {
      type: DataTypes.DATEONLY,
      allowNull: true,
    },
    asiento_numero: {
      type: DataTypes.INTEGER.UNSIGNED,
      allowNull: true,
    },
    estado: {
      type: DataTypes.STRING(255),
      allowNull: true,
      defaultValue: 'reservado',
      validate: { isIn: [ESTADO] },
    },
    precio: {
      type: DataTypes.DECIMAL(10, 2),
      allowNull: true,
    },
  },
  {
    tableName: 'Boletos',
    timestamps: true,
    underscored: true,
    indexes: [
      {
        unique: true,
        fields: ['autobus_id', 'fecha_viaje', 'asiento_numero', 'horario_id'],
      },
    ],
  }
);

// Asociaciones
Boleto.associate = (models) => {
  Boleto.belongsTo(models.Usuario, { foreignKey: 'usuarioId', as: 'usuario' });
  Boleto.belongsTo(models.Ruta, { foreignKey: 'rutaId', as: 'ruta' });
  Boleto.belongsTo(models.Autobus, { foreignKey: 'autobusId', as: 'autobus' });
  Boleto.belongsTo(models.Horario, { foreignKey: 'horarioId', as: 'horario' });
};

export default Boleto;
